File: repeat-pattern.go
package idpa

import "time"

const (
	WeekdayMonday     uint8 = 1 << time.Monday
	WeekdayTuesday    uint8 = 1 << time.Tuesday
	WeekdayWednessday uint8 = 1 << time.Wednesday
	WeekdayThursday   uint8 = 1 << time.Thursday
	WeekdayFriday     uint8 = 1 << time.Friday
	WeekdaySaturday   uint8 = 1 << time.Saturday
	WeekdaySunday     uint8 = 1 << time.Sunday
	WeekdayAll        uint8 = (1 << 7) - 1
)

const (
	MonthAll  uint16 = (1 << 12) - 1
	DayAll    uint32 = (1 << 31) - 1
	HourAll   uint32 = (1 << 24) - 1
	MinuteAll uint64 = (1 << 60) - 1
)

type RepeatPattern struct {
	MonthFlags   uint16 `json:"monthFlags"`
	DayFlags     uint32 `json:"dayFlags"`
	HourFlags    uint32 `json:"hourFlags"`
	MinuteFlags  uint64 `json:"minuteFlags"`
	WeekdayFlags uint8  `json:"weekdayFlags"`
}

func (rp RepeatPattern) Matches(t time.Time) bool {
	var (
		monthBit   uint16
		dayBit     uint32
		hourBit    uint32
		minuteBit  uint64
		weekdayBit uint8
	)

	monthBit = 1 << (t.Month() - 1)
	if rp.MonthFlags&monthBit == 0 {
		return false
	}

	dayBit = 1 << (t.Day() - 1)
	if rp.DayFlags&dayBit == 0 {
		return false
	}

	hourBit = 1 << t.Hour()
	if rp.HourFlags&hourBit == 0 {
		return false
	}

	minuteBit = 1 << t.Minute()
	if rp.MinuteFlags&minuteBit == 0 {
		return false
	}

	weekdayBit = 1 << t.Weekday()
	return rp.WeekdayFlags&weekdayBit != 0

}

File: tmp/repeatpattern.ts
interface RepeatPattern {
    monthFlags: number
    dayFlags: number
    hourFlags: number
    minuteFlags: number
    weekdayFlags: number
}

const Sunday = 0
const Monday = 1
const Tuesday = 2
const Wednesday = 3
const Thursday = 4
const Friday = 5
const Saturday = 6

function createRepeatPattern(months: number[], days: number[], hours: number[], minutes: number[], weekdays: number[]): RepeatPattern {
    let monthFlags = 0
    let dayFlags = 0
    let hourFlags = 0
    let minuteFlags = 0
    let weekdayFlags = 0

    months.forEach(m => monthFlags = monthFlags | (1 << (m - 1)))
    days.forEach(m => dayFlags = dayFlags | (1 << (m - 1)))
    hours.forEach(m => hourFlags = hourFlags | (1 << m))
    minutes.forEach(m => minuteFlags = minuteFlags | (1 << m))
    weekdays.forEach(m => weekdayFlags = weekdayFlags | (1 << m))

    return { monthFlags, dayFlags, hourFlags, minuteFlags, weekdayFlags }
}